### PR TITLE
Add utoronto staging hub

### PR DESCRIPTION
**Pull request by GeorgianaElena**
Closes https://github.com/2i2c-org/infrastructure/issues/867

### CODE REVIEW

File: config/hubs/utoronto.cluster.yaml
Comment (GeorgianaElena) [inline config/hubs/utoronto.cluster.yaml:73]: These are some resource limits from the old utoronto config. I wasn't sure whether to still use those or go with the default values in `basehub` template.

Comment (choldgraf) [reply config/hubs/utoronto.cluster.yaml:73]: IMO we should stick with the current UToronto configuration values

**Review by choldgraf — COMMENTED** [review]
This looks awesome! A couple questions to clarify:

- Now that #841 is done will this auto-deploy when we merge?
- This probably doesn't quite close #853 , because we'll need to deploy a `prod` hub as well, right?

File: config/hubs/utoronto.cluster.yaml
Comment (yuvipanda) [inline config/hubs/utoronto.cluster.yaml:19]: DNS change needs to be done for this

**Review by yuvipanda — APPROVED** [review]
This lgtm, @GeorgianaElena. You'll have to make DNS changes for staging.utoronto.2i2c.cloud to point to the IP of the nginx-ingress for this to work. But once that's done, hopefully everything (including auth!) should work.

Do deal with DNS issues, my suggested workflow is:

1. Run `deploy-support` locally to deploy nginx-ingress
2. Change DNS to point domains to the nginx-ingress service's external IP
3. Then deploy the hub after 5-10 minutes, to allow time for the DNS change to propogate

This should avoid problems with let's encrypt trying to get HTTPS working properly.

File: config/hubs/utoronto.cluster.yaml
Comment (damianavila) [reply config/hubs/utoronto.cluster.yaml:73]: Mmm... can I suggest otherwise, maybe? 😉 
I mean, do we know if these values are the proper ones for the current incantation of our hubs from this repo?
If we set the values without validation (and only because we inherit those values), we may be setting conditions that are not actually needed (nor optimal) and makes the whole thing more complex than it should be...
Maybe we can trace back why these values were needed and infer if we need to tweak them here as well before actually doing it?

**Comment by damianavila** [comment]
> Now that #841 is done will this auto-deploy when we merge?

AFAIK, nop, this one is using a raw kubeconfig and auto-deploy needs setting up some Service Principal. AFAIK, the terraform stuff for that was merged quite recently, so I am not sure the cluster supporting this hub actually was built before or after that PR was merged. In addition, I think there are still some permission issues to be figured out yet accordingly to this comment: https://github.com/2i2c-org/infrastructure/pull/864#issuecomment-984463972

**Comment by GeorgianaElena** [comment]
> Now that #841 is done will this auto-deploy when we merge?

@damianavila  is right, #841 won't work for the utoronto hub because of not enough permissions to create a service principal (https://github.com/2i2c-org/infrastructure/pull/864#pullrequestreview-820146574) :( Also, the cluster was built before the PR
 
> This probably doesn't quite close #853 , because we'll need to deploy a prod hub as well, right? Since nobody will be using this hub yet anyway, why don't we deploy both staging and prod in this PR?

I believe we should deploy the prod hub after we make sure that everything works on this staging hub. It should be a fairly straightforward task, that can happen as part of Step 2 of https://github.com/2i2c-org/infrastructure/issues/638. Also, we might need to take over the prod hub URL too (like we are doing here with the staging one).

File: config/hubs/utoronto.cluster.yaml
Comment (GeorgianaElena) [reply config/hubs/utoronto.cluster.yaml:73]: That was my feeling too @damianavila, esp since I believe the utoronto hub was the first version of our infra.

> Maybe we can trace back why these values were needed and infer if we need to tweak them here as well before actually doing it?

I will dig into it to try find out if these were tweaked afterwards or were part of the initial deployment.

**Comment by GeorgianaElena** [comment]
Ah, oops! I just noticed that in the original issue https://github.com/2i2c-org/infrastructure/issues/638, there this piece of context from @yuvipanda:

*We now have Azure terraform support in this repo (#800) and can set up a new hub infra with that, to run in parallel with current setup. The primary infra difference will be the use of Azure File for home directories rather than the hand-spun NFS VM we have there right now - I think this is a big positive change.*

I believe that NFS is enabled by default rather than Azure File, so that needs to change before merging this.

File: config/hubs/utoronto.cluster.yaml
Comment (choldgraf) [reply config/hubs/utoronto.cluster.yaml:73]: Ah - I was making an assumption here, let me re-phrase my suggestion:

- **If they have explicitly requested those limits** then we should keep them.
- **If they have not and this was just set by us without explicit rationale for special-casing it** then we should just go with whatever our other configurations use

Comment (GeorgianaElena) [reply config/hubs/utoronto.cluster.yaml:73]: Looking at the commit history, it seems that the limits were set by us.
But I'm not 100% sure this means that there wasn't a request from utoronto or a limitation observed at some point. 

l would love if @yuvipanda could confirm this :eyes: 

This is a side-by-side resource differences. Hope this makes it more easier to follow
resource-diffs

**Comment by GeorgianaElena** [comment]
I would like to move this forward since we have a migration due date.
I was thinking about the following plan:

1. Deploy the support chart for the utoronto cluster per [@yuvipanda's comment](https://github.com/2i2c-org/infrastructure/pull/866#pullrequestreview-822467441).
2. If everything goes ok in step 1, merge this PR and check if the deploy goes well.
3. Open an issue about the differences in resource allocation that we have for the utoronto hub and discuss there if we should stick with those or try use the defaults we have in the infra repo

**Comment by GeorgianaElena** [comment]
Since I have Yuvi's approval and we can tweak resources later, I will try a merge and check how the deploy goes :crossed_fingers:

**Comment by damianavila** [comment]
> Open an issue about the differences in resource allocation that we have for the utoronto hub and discuss there if we should stick with those or try use the defaults we have in the infra repo

Totally OK with me (but let's not forget about that!! 😉  )

**Comment by damianavila** [comment]
Btw, this is great progress @GeorgianaElena, thanks for pushing forward on it!!

**Comment by GeorgianaElena** [comment]
> Totally OK with me (but let's not forget about that discussion!! )

Absolutely, @damianavila!

**Pull request merged**
